xdsl/traits.py

```diff
@@ -444,30 +445,111 @@ def get_canonicalization_patterns(cls) -> tuple[RewritePattern, ...]:
+    """
+    The kind of side effect an operation can have.
+
+    MLIR has a more detailed version of this, able to tie effects to specfic resources or
+    values. Here, everything has its effect on the universe.
+    """
+
+    READ = auto()
+    """
+    Indicates that the operation reads from some resource. A 'read' effect implies only
+    dereferencing of the resource, and not any visible mutation.
+    """
+
+    WRITE = auto()
+    """
+    Indicates that the operation writes to some resource. A 'write' effect implies only
+    mutating a resource, and not any visible dereference or read.
+    """
+    ALLOC = auto()
+    """
+    Indicates that the operation allocates from some resource. An 'allocate' effect
+    implies only allocation of the resource, and not any visible mutation or dereference.
+    """
+    FREE = auto()
+    """
+    Indicates that the operation frees some resource that has been allocated. A 'free'
+    effect implies only de-allocation of the resource, and not any visible allocation,
+    mutation or dereference.
+    """
+
+
 class MemoryEffect(OpTrait):
     """
     A trait that enables operations to expose their side-effects or absence thereof.
-
-    NB: The MLIR implementation further allows to describe what *kind* of side-effects
-    an operation has, e.g., read-only, or allocation.
-    This one is a stripped down version for now, just saying if there are any
-    side-effects or not.
     """
 
     @classmethod
     @abc.abstractmethod
-    def has_effects(cls, op: Operation) -> bool:
+    def get_effects(cls, op: Operation) -> set[MemoryEffectKind]:
+        """
+        Returns the concrete side effects of the operation.
+        """
         raise NotImplementedError()
 
+    @final
+    @classmethod
+    def has_effects(cls, op: Operation) -> bool:
+        """
+        Returns if the operation has any side effects.
+        """
+        return len(cls.get_effects(op)) > 0
 
-def is_side_effect_free(op: Operation):
+    @final
+    @classmethod
+    def only_has_effect(cls, op: Operation, effect: MemoryEffectKind) -> bool:
+        """
+        Returns if the operation has the given side effects and no others.
+        """
+        return cls.get_effects(op) == {effect}
+
+
+def is_side_effect_free(op: Operation) -> bool:
     """
     Boilerplate helper to check if a generic operation is side effect free for sure.
     """
-    # If it doesn't say, safely assume it has side effects.
-    if not (trait := op.get_trait(MemoryEffect)):
+    if trait := op.get_trait(MemoryEffect):
+        # If the op implements the trait and says it has effects, done
+        if trait.has_effects(op):
+            return False
+        # If it said it has no effect, and does not have recursive effects, done
+        if not op.get_trait(RecursiveMemoryEffect):
+            return True
+    # If it implements neither trait, play safe and assume it has effects
+    elif not op.get_trait(RecursiveMemoryEffect):
```

Comment:
nit: I would swap around this elif and the fallthrough, or add a final else, to make it a tiny bit clearer that the deafault behaviour is `return False`, and the only way to return True is with one of the two effect traits

Reply:
Lifted from MLIR, I like to keep it close to what they do, because I'm only porting subsets and it makes it easier to filter out cases and add them when I add the necessary boilerplate.

Do you mind if we keep it as is until it is at least on par? Again, I'm just porting stuff, not exploring ideas on *how* best to do it or what not.

Reply:
this is fine!